### ISSUE TITLE
Support for custom certificates in TLS handler

### ISSUE THREAD

**Issue by rcm-25** [mholt/caddy-l4#244]
It would be nice to have the option to configure custom certificates for the L4 TLS handler via the Caddyfile as it is possible in Caddy with the [`tls <cert_file> <key_file>` directive](https://caddyserver.com/docs/caddyfile/directives/tls).

I would e.g. like to use it like this (in the context of docker compose):
```
{
    layer4 {
        127.0.0.1:8883 {
            route {
                tls my_cert.crt my_cert.key
                proxy <docker_container>:1883
            }
        }
    }
}
```

This feature request emerged from the discussion in [Custom certificates for TLS termination in layer4 app](https://caddy.community/t/custom-certificates-for-tls-termination-in-layer4-app/25483)

**Comment by mholt** [comment]
Good idea, (sort of, lol) -- we like automation around these parts. Why do you need manually-managed certs?

**Comment by rcm-25** [comment]
In my use case, I use Caddy with the Layer4 module as a reverse proxy in an environment that may not have online access.
I would therefore like to be able to specify my own certificate here.

**Comment by Rizary** [comment]
@rcm-25 what is your solution on this? using automatic tls?

**Comment by polarathene** [comment]
@Rizary 

Current solution/advice is to define a site-block for HTTPS if you rely on `Caddyfile` definition instead of JSON (_which allows to manage certs differently_). That will trigger cert provisioning, your site-block can have the `abort` directive if it shouldn't respond to HTTPS.

When you load a certificate from an external cert + key via the `tls` directive in a site block it'd also make it available to Caddy's cert store, which L4 `tls sni ...` can then match against for `route` (_before it's own `tls` termination_).

---

Since the current feature request here is about externally loading certs only, there is a bit of a disconnect from provisioning (_via ACME provider or locally generated by Caddy_). You could get `tls internal`, but there's no equivalent for the default ACME AFAIK? (_granted that's only for `auto_https` I think_)

---

> Why do you need manually-managed certs?

I use manually managed certs sometimes for reproduction examples that work offline. Either via externally loaded certs like requested or `tls internal`. The flexibility helped me identify why a DNS client (`kdig`) was not compatible with `tls internal`.

I just add a regular site-block to get certs into Caddy for L4 to use. Little awkward but works, especially with wildcard certs this minimizes the effort.

**Comment by vnxme** [comment]
I can see at least three approaches to obtain the requested functionality with what we already have in the HTTP and Layer 4 apps:

1. Let the Layer 4 app use certificates available for the HTTP app (it's also demonstrated [here](https://github.com/mholt/caddy-l4/blob/master/docs/examples/mutual_tls.md))

```caddyfile
{
    layer4 {
        tcp/:4443 {
            route {
                # the handler below will choose a relevant certificate
                # from all the certificates available for the HTTP app
                # based on the requested server name
                tls {
                    connection_policy {
                        # the ALPN below is required to allow proxying to port 80,
                        # if the client supports HTTP/2 in addition to HTTP/1.1
                        alpn http/1.1
                    }
                }
                proxy tcp/backend:80
            }
        }
    }
}

*.example.com {
    # load a custom TLS certificate in the HTTP app
    tls <cert_file> <key_file>
    respond "OK" 200
}
```

2. Use the Layer 4 app as a wrapper after TLS is terminated by the HTTP app (a variation of what is shown [here](https://github.com/mholt/caddy-l4/blob/master/docs/examples/combining_apps.md))

```caddyfile
{
    servers :443 {
        listener_wrappers {
            # the HTTP app terminates TLS with any certificates it has
            tls

            # the Layer 4 wrapper can be placed both before and after the `tls` wrapper;
            # this time we put it after to handle decrypted traffic
            layer4 {
                route {
                    proxy tcp/backend:80
                }
            }
        }

        # optionally disable HTTP/3, since we are talking about TLS, i.e. TCP, routing;
        # disable HTTP/2, since we aren't sure the backend can handle it properly
        protocols h1
    }
}

*.example.com {
    # load a custom TLS certificate in the HTTP app
    tls <cert_file> <key_file>
    # the response below will never be returned,
    # since all decrypted traffic is proxied to the backend
    # by the Layer 4 wrapper above
    respond "OK" 200
}
```

3. Use the HTTP app only, if the backend is HTTP/HTTPS

```caddyfile
*.example.com {
    # load a custom TLS certificate
    tls <cert_file> <key_file>
    # see also: https://caddyserver.com/docs/caddyfile/directives/reverse_proxy
    reverse_proxy backend:80
}
```

**Comment by vnxme** [comment]
Closing this issue as not planned. Feel free to reopen it or start a new discussion.
